### PR TITLE
Introduce SSL support for HTTP connections

### PR DESCRIPTION
**Pull request by mxm**
Note that this is still missing an integration test for the HBA integration. Only covered by a unit tests now. It appears there is generally no integration test for the HTTP HBA implementation yet.

### CODE REVIEW

File: users/src/main/java/io/crate/operation/auth/HostBasedAuthentication.java [hunk 1]
```diff
@@ -36,11 +36,11 @@
 
 public class HostBasedAuthentication implements Authentication {
 
-    private static final String DEFAULT_AUTH_METHOD = "trust";
-    private static final String KEY_USER = "user";
-    private static final String KEY_ADDRESS = "address";
-    private static final String KEY_METHOD = "method";
-    private static final String KEY_PROTOCOL = "protocol";
+    public static final String DEFAULT_AUTH_METHOD = "trust";
```
Comment (matriv) [inline users/src/main/java/io/crate/operation/auth/HostBasedAuthentication.java]: is it necessary to make them public?

File: ssl/src/main/java/io/crate/protocols/ssl/SslHandlerLoader.java [hunk 2]
```diff
@@ -20,38 +20,63 @@
...
 
     /**
      * Loads the SslRequest handler. Should only be called once per application life time.
      * @throws SslConfigurationException Exception thrown if the user has supplied an invalid configuration
      */
-    public static SslReqHandler load(Settings settings) {
+    public static SslReqHandler loadSslReqHandler(Settings settings) {
+        return load(settings,
+            SslConfigSettings.SSL_PSQL_ENABLED,
+            "io.crate.protocols.postgres.SslReqConfiguringHandler",
```
Comment (matriv) [inline ssl/src/main/java/io/crate/protocols/ssl/SslHandlerLoader.java]: I would keep the fully qualified class name in a static final var.

File: ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java [hunk 3]
```diff
@@ -179,15 +179,23 @@ public void testExportServerCertChain() throws Exception {
...
-        assertThat(x509Certificates[0].getSubjectDN().getName(), containsString("CN=CrateDB"));
-        assertThat(x509Certificates[0].getSubjectDN().getName(), containsString("PSQL Netty SSL"));
-        assertThat(x509Certificates[0].getSubjectDN().getName(), containsString("L=Dornbirn"));
-        assertThat(x509Certificates[1].getSubjectDN().getName(), containsString("CN=crate.io"));
+        assertThat(x509Certificates.length, is(4));
+        assertThat(x509Certificates[0].getIssuerDN().getName(), containsString("CN=*.crate.io"));
+        assertThat(x509Certificates[1].getIssuerDN().getName(), containsString("CN=*.crate.io"));
+        assertThat(x509Certificates[2].getIssuerDN().getName(), containsString("CN=*.crate.io"));
+        assertThat(x509Certificates[3].getIssuerDN().getName(), containsString("CN=*.crate.io"));
+        assertThat(x509Certificates[0].getSubjectDN().getName(), containsString("CN=localhost"));
```
Comment (matriv) [inline ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java]: Can't we test this in one line and not with 3 different containsString() assertions?

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java [hunk 4]
```diff
@@ -84,7 +85,7 @@ static SslContext buildSslContext(Settings settings) {
                 trustManagers = trustStoreSettings.get().trustManagers;
             }
 
-            SSLContext sslContext = SSLContext.getInstance("TLS");
+            SSLContext sslContext = SSLContext.getInstance("TLSv1.2");
```
Comment (matriv) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: Maybe add a comment why we need that (the browser problem that you mentioned I guess).

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java [hunk 5]
```diff
@@ -107,6 +107,7 @@ static SslContext buildSslContext(Settings settings) {
                     .clientAuth(ClientAuth.OPTIONAL)
                     .sessionCacheSize(0)
                     .sessionTimeout(0)
+                    .startTls(false)
```
Comment (matriv) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: is this still required even with the specific TLSv1.2 specified?

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java [hunk 6]
```diff
@@ -178,6 +179,8 @@ static SslContext buildSslContext(Settings settings) {
         }
 
         X509Certificate[] exportServerCertChain() throws KeyStoreException {
+
```
Comment (matriv) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: empty line :-P

File: blackbox/docs/administration/ssl.txt [hunk 7]
```diff
@@ -128,29 +170,31 @@ to get you started. All the commands use a validity of 36500 days
...
-    keytool -keystore keystore -genkey -alias server -validity 36500
+    keytool -keystore keystore -genkey -keyalg RSA -alias server -validity 36500
 
 Output::
 
     Enter keystore password:
     Re-enter new password:
     What is your first and last name?
-      [Unknown]:  CrateDB
+      [Unknown]:  ssl.crate.io
```
Comment (matriv) [inline blackbox/docs/administration/ssl.txt]: check for extra empty space after `:` in this section.

File: blackbox/docs/administration/ssl.txt [hunk 8]
```diff
@@ -195,10 +239,13 @@ Generate a Self-Signed Certificate
 
 If you don't get your certificate signed from one of the official CAs,
 you might want to create your own CA with a self-signed certificate.
+The common name (CN) should overlap with the CN of the server key
```
Comment (matriv) [inline blackbox/docs/administration/ssl.txt]: Sorry, I don't understand this sentence.

File: blackbox/docs/configuration.txt [hunk 9]
```diff
@@ -1901,12 +1901,19 @@ Secured communications via SSL allows you to encrypt traffic between CrateDB
...
+  Set this to true to enable secure communication between the CrateDB node
+  and the client through SSL via the HTTPS protocol.
+
 **ssl.psql.enabled**
   | *Runtime:*  ``no``
   | *Default:* ``false``
 
-  Set this to true to enable secured communications between CrateDB node and
-  client via SSL via PostgreSQL wire protocol.
+  Set this to true to enable secure communication between tbe CrateDB node
```
Comment (matriv) [inline blackbox/docs/configuration.txt]: tbe => the

File: blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java [hunk 10]
```diff
@@ -367,22 +380,35 @@ private void fullContentResponse(HttpRequest request, String index, final String
...
 
-        final FileRegion region = new DefaultFileRegion(raf.getChannel(), position, count);
         Channel channel = ctx.channel();
-        ChannelFuture writeFuture = channel.write(region, ctx.newProgressivePromise());
-        writeFuture.addListener(new ChannelProgressiveFutureListener() {
+        final ChannelFuture fileFuture;
+        final ChannelFuture endMarkerFuture;
+        if (sslEnabled) {
+            HttpChunkedInput httpChunkedInput =
+                new HttpChunkedInput(new ChunkedFile(raf, 0, count, 8192));
```
Comment (matriv) [inline blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java]: extract 8192 to a constant. Is this something we need to benchmark and tune?

**Comment by matriv** [comment]
Maybe add an example with http and ssl (on or off) here: https://github.com/crate/crate/blob/master/blackbox/docs/administration/hba.txt ?

File: http/src/main/java/io/crate/plugin/PipelineRegistry.java [hunk 11]
```diff
@@ -73,6 +92,7 @@ public void registerItems(ChannelPipeline pipeline) {
         for (PipelineRegistry.ChannelPipelineItem item : addBeforeList) {
             pipeline.addBefore(item.base, item.name, item.handlerFactory.get());
         }
+        sslHandler.addToPipeline(pipeline);
```
Comment (mfussenegger) [inline http/src/main/java/io/crate/plugin/PipelineRegistry.java]: Not sure i like that this is bi-directional. Wouldn't it be possible to remove any SslHandler knowledge from the PipelineRegistry? I think we should have some service in `ssl-impl` that creates the context and then simply uses the PipelineRegistry to add the SslHandler. This would also remove the need for different `HttpsHandler` implementations

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java (hunk 4 above)
Comment (mfussenegger) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: Would it make sense to extract the changes here to a separate PR against master? Seem to be independent of HTTPS

File: users/src/main/java/io/crate/operation/auth/HostBasedAuthentication.java (hunk 1 above)
Comment (mxm) [inline users/src/main/java/io/crate/operation/auth/HostBasedAuthentication.java]: I think they should be public because these are hardcoded in a lot of places which makes it difficult to change them. I avoided to hardcode in my tests. Generally, `static final` should be safe to be made public.

File: ssl/src/main/java/io/crate/protocols/ssl/SslHandlerLoader.java (hunk 2 above)
Comment (mxm) [inline ssl/src/main/java/io/crate/protocols/ssl/SslHandlerLoader.java]: I was wondering whether this should be exposed at all. Since there are multiple messages now, I wanted to ensure the classes don't get mixed up. However, there are type checks in place, so should be safe.

File: ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java (hunk 3 above)
Comment (mxm) [inline ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java]: Could you explain the benefit of doing that? I think three lines give a better error reporting and are easier to read.

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java (hunk 4 above)
Comment (mxm) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: This is the newest standard of TLS. If you get "TLS", the TLS version depends on the JVM you're using. I wanted to be explicit here. Will add a comment.

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java (hunk 5 above)
Comment (mxm) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: Not required but more explicit since we don't use StartTLS. This is not dependent on the TLS version. IMHO it doesn't hurt.

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java (hunk 6 above)
Comment (mxm) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: 👅

File: blackbox/docs/administration/ssl.txt (hunk 7 above)
Comment (mxm) [inline blackbox/docs/administration/ssl.txt]: The output has two spaces are the colon. That's how the keytool outputs.

File: blackbox/docs/administration/ssl.txt (hunk 8 above)
Comment (mxm) [inline blackbox/docs/administration/ssl.txt]: You're right. Should be clarified. What I mean is that `*.crate.io` overlaps `ssl.crate.io`.

File: blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java (hunk 10 above)
Comment (mxm) [inline blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java]: Will investigate if tuning makes sense here.

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java (hunk 4 above)
Comment (mxm) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: I was thinking about opening a separate PR. I can do that if you want. Actually, it is not completely unrelated since we depend on a recent TLS version to support all browsers.

File: ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java (hunk 3 above)
Comment (matriv) [inline ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java]: My thought is that with containsString you don't test the strict equality of the cert string after all

File: ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java (hunk 4 above)
Comment (matriv) [inline ssl-impl/src/main/java/io/crate/protocols/ssl/SslConfiguration.java]: Thx! Just for future reminder

File: http/src/main/java/io/crate/plugin/PipelineRegistry.java (hunk 11 above)
Comment (mxm) [inline http/src/main/java/io/crate/plugin/PipelineRegistry.java]: Fair point. I think we can remove the default implementation and pass the PipelineRegistry via the constructor arg of the ssl implementation.

**Comment by mxm** [comment]
> Also, are you using different keystore/trustore .jks files for the tests, could we symlink them?

AFAIK I'm just using one version.

>Maybe add an example with http and ssl (on or off) here: https://github.com/crate/crate/blob/master/blackbox/docs/administration/hba.txt ?

👍

File: ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java (hunk 3 above)
Comment (mxm) [inline ssl-impl/src/test/java/io/crate/protocols/ssl/SslConfigurationTest.java]: This is just the name, not the cert itself. After all, this is not to a security check here. It just assures the all certificates have been loaded.

File: blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java (hunk 10 above)
Comment (mxm) [inline blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java]: I think it is hard to tell what the optimal size is because it depends on the network environment. I think having larger chunks generally is beneficial for the download speed. We can revisit when we do more benchmarks. If necessary, we can make the default configurable.

**Comment by mxm** [comment]
Thank you for your comments @matriv and @mfussenegger. I moved some of the commits to separate PRs: #5767 and #5768. Also, I addressed your comments.

File: blackbox/docs/administration/hba.txt [hunk 12]
```diff
@@ -119,9 +119,9 @@ using the ``trust`` method.
 that the user ``dustin`` can authenticate to CrateDB over HTTP protocol from
 the ``172.16.0.0`` IP Address using the ``trust`` method.
 
-``{user: trinity, address: 127.0.0.1, protocol: pg, ssl: off}`` means that the
-user ``trinity`` can authenticate to CrateDB over Postgres Wire Protocol from
-the ``127.0.0.1`` IP Address only if no SSL/TLS connection is used. Since no
+``{user: trinity, address: 127.0.0.1, protocol: http, ssl: off}`` means that
```
Comment (matriv) [inline blackbox/docs/administration/hba.txt]: the example above still says protocol: pg, please change
https://github.com/crate/crate/blob/8d286dfa88b2d3ac5e30fd70a8c3e6fb708167e6/blackbox/docs/administration/hba.txt#L81

File: blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java (hunk 10 above)
Comment (matriv) [inline blob/src/main/java/io/crate/protocols/http/HttpBlobHandler.java]: I would still extract it in a static final var to make it more visible.

File: blackbox/docs/administration/hba.txt (hunk 12 above)
Comment (mxm) [inline blackbox/docs/administration/hba.txt]: Good catch!
